src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs

```diff
@@ -109,8 +108,8 @@ internal void OnIncompleteOperation(int expectedBytesTransferred, int actualByte
         /// <summary>Gets or sets the position within the current stream</summary>
         public sealed override long Position
         {
-            get => _filePosition;
-            set => _filePosition = value;
+            get => Interlocked.Read(ref _filePosition);
+            set => Interlocked.Exchange(ref _filePosition, value);
```

Comment:
How can this have an outdated value? The failing test (`WriteAsyncCancelledFile`) is creating a new file and a new `FileStream` instance in each run, right?

Reply:
It might be caused by false sharing, but I am just guessing: https://docs.microsoft.com/en-us/archive/msdn-magazine/2008/october/net-matters-false-sharing

Reply:
Wow, that's a good article.  The lead author must really know what he's talking about. :smile:

That said, this isn't false sharing.  @Jozkee is correct that this value shouldn't be outdated, as it comes linearly after the operations should have completed and quiesced with appropriate synchronization to ensure that all updates were published.  I don't know why it's failing, nor why this fixes it, but it suggests a real problem with something potentially being done in a problematic order in the code, such as accidentally changing _filePosition after completing the task.